### PR TITLE
feat: support matrices and stdout for all output types

### PR DESCRIPTION
**Pull request by trxcllnt**
* Support `stdout` for all output types via the new `--stdout` flag
* Allow passing `--file_key`, `--output`, or `--matrix` (or any combination of the three)
* Prepends the comment to the header for all file types
* Support multiple `--file_key` and `--output` arguments

Fixes https://github.com/rapidsai/dependency-file-generator/issues/46

Enables the `dependencies.yaml` changes in [this branch](https://github.com/trxcllnt/cudf/tree/fix/generated-pip-dependencies).

<details><summary>Example generating requirements.txt:</summary><pre>$ rapids-dependency-file-generator --output requirements -f py_build_cudf -f py_run_cudf --matrix &quot;cuda=11.8;arch=$(uname -m)&quot; --stdout<br/># This file is generated by `rapids-dependency-file-generator`.<br/># To make changes, edit ../../dependencies.yaml and run `rapids-dependency-file-generator`.<br/>cmake&gt;=3.23.1,!=3.25.0<br/>cython&gt;=0.29,&lt;0.30<br/>git+https://github.com/python-streamz/streamz.git@master<br/>ninja<br/>numpy&gt;=1.21,&lt;1.24<br/>protoc-wheel<br/>pyarrow==11.0.0.*<br/>rmm-cu11==23.6.*<br/>scikit-build&gt;=0.13.1,&lt;0.17.2<br/># This file is generated by `rapids-dependency-file-generator`.<br/># To make changes, edit ../../dependencies.yaml and run `rapids-dependency-file-generator`.<br/>--extra-index-url=https://pypi.nvidia.com<br/>cachetools<br/>cubinlinker-cu11<br/>cuda-python&gt;=11.7.1,&lt;12.0<br/>cupy-cuda11x&gt;=12.0.0<br/>fsspec&gt;=0.6.0<br/>numba&gt;=0.56.4,&lt;0.57<br/>numpy&gt;=1.21,&lt;1.24<br/>nvtx&gt;=0.2.1<br/>packaging<br/>pandas&gt;=1.3,&lt;1.6.0dev0<br/>protobuf&gt;=4.21.6,&lt;4.22<br/>ptxcompiler-cu11<br/>pyarrow==11.*<br/>rmm-cu11==23.6.*<br/>typing_extensions<br/></pre></details>

<details><summary>Example generating pyproject.toml</summary><pre>$ rapids-dependency-file-generator --output pyproject -f py_build_cudf -f py_run_cudf --matrix &quot;cuda=11.8;arch=$(uname -m)&quot; --stdout</br># This file is generated by `rapids-dependency-file-generator`.</br># To make changes, edit ../../dependencies.yaml and run `rapids-dependency-file-generator`.</br># Copyright (c) 2021-2023, NVIDIA CORPORATION.</br></br>[build-system]</br>build-backend = &quot;setuptools.build_meta&quot;</br>requires = [</br>    &quot;cmake&gt;=3.23.1,!=3.25.0&quot;,</br>    &quot;cython&gt;=0.29,&lt;0.30&quot;,</br>    &quot;ninja&quot;,</br>    &quot;numpy&gt;=1.21,&lt;1.24&quot;,</br>    &quot;protoc-wheel&quot;,</br>    &quot;pyarrow==11.0.0.*&quot;,</br>    &quot;rmm-cu11==23.6.*&quot;,</br>    &quot;scikit-build&gt;=0.13.1,&lt;0.17.2&quot;,</br>    &quot;setuptools&quot;,</br>    &quot;wheel&quot;,</br>]</br></br>[project]</br>name = &quot;cudf-cu11&quot;</br>version = &quot;23.6.0&quot;</br>description = &quot;cuDF - GPU Dataframe&quot;</br>readme = { file = &quot;README.md&quot;, content-type = &quot;text/markdown&quot; }</br>authors = [</br>    { name = &quot;NVIDIA Corporation&quot; },</br>]</br>license = { text = &quot;Apache 2.0&quot; }</br>requires-python = &quot;&gt;=3.9&quot;</br>dependencies = [</br>    &quot;cachetools&quot;,</br>    &quot;cubinlinker-cu11&quot;,</br>    &quot;cuda-python&gt;=11.7.1,&lt;12.0&quot;,</br>    &quot;cupy-cuda11x&gt;=12.0.0&quot;,</br>    &quot;fsspec&gt;=0.6.0&quot;,</br>    &quot;numba&gt;=0.56.4,&lt;0.57&quot;,</br>    &quot;numpy&gt;=1.21,&lt;1.24&quot;,</br>    &quot;nvtx&gt;=0.2.1&quot;,</br>    &quot;packaging&quot;,</br>    &quot;pandas&gt;=1.3,&lt;1.6.0dev0&quot;,</br>    &quot;protobuf&gt;=4.21.6,&lt;4.22&quot;,</br>    &quot;ptxcompiler-cu11&quot;,</br>    &quot;pyarrow==11.*&quot;,</br>    &quot;rmm-cu11==23.6.*&quot;,</br>    &quot;typing_extensions&quot;,</br>]</br>classifiers = [</br>    &quot;Intended Audience :: Developers&quot;,</br>    &quot;Topic :: Database&quot;,</br>    &quot;Topic :: Scientific/Engineering&quot;,</br>    &quot;License :: OSI Approved :: Apache Software License&quot;,</br>    &quot;Programming Language :: Python&quot;,</br>    &quot;Programming Language :: Python :: 3.9&quot;,</br>    &quot;Programming Language :: Python :: 3.10&quot;,</br>]</br></br>[project.optional-dependencies]</br>test = [</br>    &quot;fastavro&gt;=0.22.9&quot;,</br>    &quot;hypothesis&quot;,</br>    &quot;mimesis&gt;=4.1.0&quot;,</br>    &quot;msgpack&quot;,</br>    &quot;pyorc&quot;,</br>    &quot;pytest&quot;,</br>    &quot;pytest-benchmark&quot;,</br>    &quot;pytest-cases&quot;,</br>    &quot;pytest-cov&quot;,</br>    &quot;pytest-xdist&quot;,</br>    &quot;python-snappy&gt;=0.6.0&quot;,</br>    &quot;scipy&quot;,</br>    &quot;tokenizers==0.13.1&quot;,</br>    &quot;transformers==4.24.0&quot;,</br>    &quot;tzdata&quot;,</br>]</br></br>[project.urls]</br>Homepage = &quot;https://github.com/rapidsai/cudf&quot;</br>Documentation = &quot;https://docs.rapids.ai/api/cudf/stable/&quot;</br></br>[tool.setuptools]</br>license-files = [&quot;LICENSE&quot;]</br></br>[tool.isort]</br>line_length = 79</br>multi_line_output = 3</br>include_trailing_comma = true</br>force_grid_wrap = 0</br>combine_as_imports = true</br>order_by_type = true</br>known_dask = [</br>    &quot;dask&quot;,</br>    &quot;distributed&quot;,</br>    &quot;dask_cuda&quot;,</br>]</br>known_rapids = [</br>    &quot;rmm&quot;,</br>]</br>known_first_party = [</br>    &quot;cudf&quot;,</br>]</br>default_section = &quot;THIRDPARTY&quot;</br>sections = [</br>    &quot;FUTURE&quot;,</br>    &quot;STDLIB&quot;,</br>    &quot;THIRDPARTY&quot;,</br>    &quot;DASK&quot;,</br>    &quot;RAPIDS&quot;,</br>    &quot;FIRSTPARTY&quot;,</br>    &quot;LOCALFOLDER&quot;,</br>]</br>skip = [</br>    &quot;thirdparty&quot;,</br>    &quot;.eggs&quot;,</br>    &quot;.git&quot;,</br>    &quot;.hg&quot;,</br>    &quot;.mypy_cache&quot;,</br>    &quot;.tox&quot;,</br>    &quot;.venv&quot;,</br>    &quot;_build&quot;,</br>    &quot;buck-out&quot;,</br>    &quot;build&quot;,</br>    &quot;dist&quot;,</br>    &quot;__init__.py&quot;,</br>]</br></pre></details>

### CODE REVIEW

**Comment by ajschmidt8** [comment]
I think these changes are considered breaking because of the new `stdout` flag/behavior.

It will break test scripts like this: https://github.com/rapidsai/cudf/blob/72c067726ccfb6e87033d34ab07b4dc79b5e4a3e/ci/test_python_common.sh#L10-L14

@trxcllnt, can you add a `BREAKING CHANGE` note to your PR body as mentioned here https://github.com/rapidsai/dependency-file-generator/blob/main/CONTRIBUTING.md? That will make sure the next release increments the major version.

Our CI images pin to the current major version of `dfg`: https://github.com/rapidsai/ci-imgs/blob/cac1028880574b466ed37a4aec8aaf93d3eab0b2/Dockerfile#L117-L119

But we'll need a way to incrementally roll this out to each repository before we update the version in our CI images so that we don't break CI for everyone.

I thought about this in the past, but never took any action on it due to time constraints.

I think one way we can fix it is to add a new optional input, `install_dfg_version`, to the relevant shared workflows here: https://github.com/rapidsai/shared-action-workflows/tree/branch-23.06/.github/workflows.

`install_dfg_version` will be a version specifier for `rapids-dependency-file-generator`. When it's set, it will install the specified version. When it is not set, it will simply do nothing.

Then each repo can manually opt-in to the new major version. Once all the repos are using the new version, we can update the CI image version accordingly and then go back and clean up all of the optional `install_dfg_version` arguments in each repository.

File: src/rapids_dependency_file_generator/rapids_dependency_file_generator.py [hunk 1]
```diff
@@ -394,7 +410,7 @@ def make_dependency_files(parsed_config, config_file_path, to_stdout):
                             # exists. In that case we save the fallback_entry result
                             # and only use it at the end if nothing more
                             # specific is found.
-                            if not specific_matrices_entry["matrix"]:
+                            if not specific_matrices_entry.get("matrix", None):
```
Comment (vyasr) [inline src/rapids_dependency_file_generator/rapids_dependency_file_generator.py:413]: When is the matrix key optional? For a specific entry it should be required, even with the new CLI functionality, right?

File: src/rapids_dependency_file_generator/rapids_dependency_file_generator.py [hunk 2]
```diff
@@ -413,43 +429,78 @@ def make_dependency_files(parsed_config, config_file_path, to_stdout):
...
+                        with open(output_file_path) as f:
+                            results[output_file_path] = tomlkit.load(f)
+                            # Append `-cuXX` to `[package.name]`
+                            results[output_file_path]["project"][
+                                "name"
+                            ] = name_with_cuda_suffix(
+                                results[output_file_path]["project"]["name"],
+                                matrix_combo.get("cuda", None),
+                                cuda_suffix,
+                            )
```
Comment (vyasr) [inline src/rapids_dependency_file_generator/rapids_dependency_file_generator.py:466]: When I first saw the function I actually thought the goal was to handle _dependency_ suffixes rather than package name suffixes. I'm even less comfortable putting this logic into this tool. Package renaming to support our wheels-specific workflows is definitely scope creep. I'd love to find a better solution to what we're currently doing in our wheel builds, but I don't think this is it.

File: src/rapids_dependency_file_generator/rapids_dependency_file_generator.py [hunk 3]
```diff
@@ -413,43 +429,78 @@ def make_dependency_files(parsed_config, config_file_path, to_stdout):
...
+
+        relpath_to_config_file = os.path.relpath(config_file_path, output_dir)
+        header = textwrap.dedent(
+            f"""\
+            {HEADER}
+            # To make changes, edit {relpath_to_config_file} and run `{cli_name}`.
+            """
+        )
+
+        if isinstance(data, dict):
```
Comment (vyasr) [inline src/rapids_dependency_file_generator/rapids_dependency_file_generator.py:487]: Is this implicitly relying on pyproject outputs generating a dict here while other outputs just contain a long string of text? We should make that condition explicit if so, otherwise it's confusing why tomlkit is getting used in a generic dict path.

File: src/rapids_dependency_file_generator/rapids_dependency_file_generator.py [hunk 4]
```diff
@@ -413,43 +429,78 @@ def make_dependency_files(parsed_config, config_file_path, to_stdout):
...
-                    os.makedirs(output_dir, exist_ok=True)
-                    file_path = os.path.join(output_dir, full_file_name)
-                    with open(file_path, "w") as f:
-                        if file_type == str(OutputTypes.PYPROJECT):
-                            tomlkit.dump(contents, f)
-                        else:
-                            f.write(contents)
+    def write_output(data, output_dir, f):
+
+        relpath_to_config_file = os.path.relpath(config_file_path, output_dir)
```
Comment (vyasr) [inline src/rapids_dependency_file_generator/rapids_dependency_file_generator.py:479]: If we hoist this logic into the calling loop then the `write_output` function could be moved outside instead of defined as a nested function right? I think that would be cleaner if we do decide to keep it.

**Review by vyasr — CHANGES_REQUESTED** [review]
Thanks for the PR! I have a couple of concerns about the current implementation but I think the core work around extending the CLI is a step in the right direction.

File: src/rapids_dependency_file_generator/rapids_dependency_file_generator.py [hunk 5]
```diff
@@ -323,7 +315,34 @@ def should_use_specific_entry(matrix_combo, specific_entry_matrix):
     )
 
 
-def make_dependency_files(parsed_config, config_file_path, to_stdout):
+def name_with_cuda_suffix(name, cuda_version=None, cuda_suffix="-cu"):
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply src/rapids_dependency_file_generator/rapids_dependency_file_generator.py:318]: We can (and do) put the `-cuXX` names in the `dependencies.yaml` matrix entries. This is explicitly about the `name` key of the `pyproject.toml`.

Overall `pyproject.toml` should probably be fully auto-generated instead of read and mutated.

File: src/rapids_dependency_file_generator/cli.py [hunk 6]
```diff
@@ -87,18 +115,29 @@ def main(argv=None):
...
+            if parsed_config["files"].get(file_key, None) is not None
         }
 
+    if len(args.output) > 0:
+        files = {}
+        for file_key, file_config in parsed_config["files"].items():
+            outputs = get_requested_output_types(file_config["output"])
+            outputs = list(filter(lambda output: output in outputs, args.output))
+            if len(outputs) > 0:
+                file_config["output"] = outputs
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply src/rapids_dependency_file_generator/cli.py:132]: This PR supports passing multiple output types, so the logic as-is is correct. I filter the outputs for ones that occur in `args.output` on L130.

File: src/rapids_dependency_file_generator/rapids_dependency_file_generator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (trxcllnt) [reply src/rapids_dependency_file_generator/rapids_dependency_file_generator.py:318]: @vyasr where do you think the logic to define the `project.name` key in the `pyproject.toml` should live?

[the remark this answers is not in the document]
Comment (vyasr) [reply src/rapids_dependency_file_generator/rapids_dependency_file_generator.py:318]: When you say fully auto-generated, what are you envisioning? I could see using something like a Jinja templated pyproject.toml.in and filling in a suitable set of fields, perhaps. There are large swathes of the file that are necessary for things like 1) running linters, 2) specifying "extra" dependency lists, 3) configuring build backends, and more that have to be encoded somewhere else. None of those should be dependency-file-generator's responsibility, and at least the first one requires the file to already exist somewhere in the repo for normal usage.

[the remark this answers is not in the document]
Comment (vyasr) [reply src/rapids_dependency_file_generator/rapids_dependency_file_generator.py:318]: Regarding the `project.name`, I honestly do not know. My approach to taking the magic out of the wheel building process so far has been very incremental, fixing one problem at a time. I've been viewing the name as sort of a final frontier, one for which I don't have a good answer for yet unfortunately. I don't think a tool dedicated to dependency management is the right place to put that, though.

File: src/rapids_dependency_file_generator/cli.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply src/rapids_dependency_file_generator/cli.py:132]: I got confused by the logic swapping which variable means what here because the filtered `args.output` is being assigned back to `outputs` after checking for membership in `outputs`. You are right, the filtering makes sense.

**Comment by vyasr** [comment]
@trxcllnt apologies for the delay in reviewing this PR. I propose that we we split this up so that the uncontroversial pieces can get merged quickly, if you're open to that.
- I think everyone is happy with adding support for `stdout` to all file types (especially pyproject.toml)
- My guess is that generalizing the CLI to support only passing a subset of arguments is probably uncontroversial in theory, but a bit more problematic because it's a breaking change. I know you said you and @ajschmidt8 had discussed that piece a bit further, and maybe there's a path forward there that involves refactoring the Python functionality and then using it in a new CLI? That would be another good option.
- The project name/CUDA suffix piece will need some more debate, and I don't want to hold up you getting the other useful pieces in over that.

**Comment by vyasr** [comment]
@trxcllnt would you like some help finishing this up? Let me know if you want to chat about it or need some extra person-hour help.

File: src/rapids_dependency_file_generator/cli.py [hunk 7]
```diff
@@ -32,39 +34,65 @@ def validate_args(argv):
...
-            "The following arguments must be used together:"
-            + "".join([f"\n  --{x}" for x in dependent_arg_keys])
-        )
 
     # If --clean was passed without arguments, default to cleaning from the root of the
     # tree where the config file is.
     if args.clean == "":
         args.clean = os.path.dirname(os.path.abspath(args.config))
 
+    args.file_key = list(sorted(list(set(sorted(args.file_key)))))
```
[the remark this answers is not in the document]
Comment (msarahan) [reply src/rapids_dependency_file_generator/cli.py:90]: Yes, it gets converted to a list:

```
>>> abc = set((1,2,3,4,5))
>>> abc
{1, 2, 3, 4, 5}
>>> sorted(abc)
[1, 2, 3, 4, 5]
```

[the remark this answers is not in the document]
Comment (vyasr) [reply src/rapids_dependency_file_generator/cli.py:90]: Right, and in general there are a lot of extra ops here.

**Comment by bdice** [comment]
@trxcllnt @vyasr Can we get this PR to a completed state? It keeps biting us that we don't have pyproject matrix support. I can help with a review, if that's what is needed, but it seems like there is some code work to be done still.

**Comment by vyasr** [comment]
Replaced by #74
